_posts/2023-01-25-auto-restarting-connectors.md

```diff
@@ -0,0 +1,317 @@
+---
+layout: post
+title:  "Auto-restarting connectors in Apache Kafka Connect and Mirror Maker 2"
+date: 2023-01-25
+author: jakub_scholz
+---
+
+Apache Kafka Connect provides a framework for integrating Apache Kafka with external systems using connectors.
+Connectors run inside the Kafka Connect deployment, connect to the external system, and push data from Kafka to the external system or the other way around.
+When running Kafka Connect, you have to monitor not only the state of the Kafka Connect deployment, but also the state of the connectors and their tasks.
+If a connector or one of its tasks fail, you check the reason it failed so you know how to fix it.
+In many cases, the issue is something temporary like a network glitch or an outage of the external system.
+And all you need to do to fix it is just restart the connector or task.
+Could this be something the operator might do for you?
+
+<!--more-->
+
+Having the operator automatically restart failed connectors or their tasks was one of the most common feature requests received by Strimzi.
+And in Strimzi 0.33, this feature was contributed by [Thomas Dangleterre](https://github.com/ThomasDangleterre).
+So, how does it work?
+
+The auto-restart feature is supported only when using the _connector operator_ and the `KafkaConnector` custom resources to manage the connectors.
+It is disabled by default.
+You can enable it using the `.spec.autoRestart` property of the `KafkaConnector` resource:
+
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: KafkaConnector
+metadata:
+  name: my-connector
+spec:
+  # ...
+  autoRestart:
+    enabled: true
+  # ...
+```
+
+When enabled, the Strimzi Cluster Operator watches the connector and its tasks for failures in every reconciliation.
+And if it sees that any of them have failed, it automatically restarts them.
+
+Not every error can be solved by a restart.
+If the connector or its task failed by some temporary problem such as the network issue mentioned earlier, a restart is an obvious solution.
+But if it is failing because it does not understand the protocol of the remote system or has the wrong credentials, then a restart will not help.
+In these cases, we do not want to keep restarting the connectors forever.
+
+That is why the operator will always restart the connector or its task only up to 7 times.
```

Comment:
Is the number of attempts configurable? If not can we make it clear?

Reply:
No, it is not. I don't think it will sound friendly if you state there that it is not configurable.